File: java/src/main/java/adventofcode/day1/ReportRepair.java
package adventofcode.day1;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

// Day 1 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/1
public class ReportRepair {

  private static final int TARGET = 2020;

  // Time complexity: O(N)
  // Space complexity: O(N)
  public static int[] solutionPart1() {
    List<Integer> input = parseInput();
    HashSet<Integer> set = new HashSet<>();

    for (int num : input) {
      int remainder = TARGET - num;
      if (set.contains(remainder)) {
        return new int[]{remainder, num};
      }
      set.add(num);
    }

    // Can't find pairs that add to 2020
    return new int[]{-1, -1};
  }

  // Time Complexity: O(N)
  // Space Complexity: O(1), assuming merge sorting in place
  public static int[] solutionPart2() {
    List<Integer> input = parseInput();
    Collections.sort(input);

    for(int i  = 0; i < input.size() - 2; i++) {
      int start = i + 1;
      int end = input.size() - 1;

      while (start < end) {
        int result = input.get(i) + input.get(start) + input.get(end);
        if (result == TARGET) {
          return new int[]{input.get(i), input.get(start), input.get(end)};
        } else if (result < TARGET) {
          start++;
        } else {
          end--;
        }
      }
    }

    // Can't find triplet
    return new int[]{-1, -1, -1};
  }

  private static List<Integer> parseInput() {
    List<Integer> result = new ArrayList<>();

    try {
      FileReader fileReader = new FileReader("java/src/main/resources/inputs/day1.txt");
      BufferedReader bufferedReader = new BufferedReader(fileReader);
      String line = bufferedReader.readLine();
      while (line != null) {
        result.add(Integer.parseInt(line));
        line = bufferedReader.readLine();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return result;
  }

  public static void main(String[] args) {
    int[] pair = solutionPart1();
    int[] triplet = solutionPart2();
    int result1 = pair[0] * pair[1];
    int result2 = triplet[0] * triplet[1] * triplet[2];
    System.out.println(String.format("Pair: [%d, %d], Result: %d", pair[0], pair[1], result1));
    System.out.println(String.format("Triplet: [%d, %d, %d], Result: %d", triplet[0], triplet[1], triplet[2], result2));
  }
}

File: golang/day1/report_repair.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

// Day 1 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/1

var TARGET int64 = 2020

func parseInput() []int64 {
	var result []int64
	inputFile, error := os.Open("input.txt")
	if error != nil {
		log.Fatal(error)
	}
	defer inputFile.Close()

	scanner := bufio.NewScanner(inputFile)
	for scanner.Scan() {
		num, _ := strconv.ParseInt(scanner.Text(), 10, 64)
		result = append(result, num)
	}

	if error := scanner.Err(); error != nil {
		log.Fatal(error)
	}

	return result
}

func solutionPart1() [2]int64 {
	var input []int64 = parseInput()
	var set map[int64]bool = make(map[int64]bool)

	for _, num := range input {
		var remainder int64 = TARGET - num
		if set[remainder] {
			return [2]int64{num, remainder}
		}

		set[num] = true
	}

	return [2]int64{-1, -1}
}

func solutionPart2() [3]int64 {
	var input []int64 = parseInput()
	sort.Slice(input, func(e1, e2 int) bool {
		return input[e1] < input[e2]
	})

	for i := 0; i < len(input)-2; i++ {
		start := i + 1
		end := len(input) - 1

		for start < end {
			sum := input[i] + input[start] + input[end]

			if sum == TARGET {
				return [3]int64{input[i], input[start], input[end]}
			} else if sum < TARGET {
				start++
			} else {
				end--
			}
		}
	}

	return [3]int64{-1, -1, -1}
}

func main() {
	var result1 [2]int64 = solutionPart1()
	var result2 [3]int64 = solutionPart2()
	fmt.Printf("Pair: [%d, %d], Result: %d", result1[0], result1[1], result1[0]*result1[1])
	fmt.Println("")
	fmt.Printf("Triplet: [%d, %d, %d], Result: %d", result2[0], result2[1], result2[2], result2[0]*result2[1]*result2[2])
}
